python/ray/data/tests/test_map.py

```diff
@@ -1564,6 +1564,29 @@ def __del__(self):
     wait_for_condition(lambda: not os.path.exists(test_file))
 
 
+def test_warn_large_udfs(ray_start_regular_shared):
+    driver = """
+import ray
+import numpy as np
+from ray.data._internal.execution.operators.map_operator import MapOperator
+
+large_object = np.zeros(MapOperator.MAP_UDF_WARN_SIZE_THRESHOLD + 1, dtype=np.int8)
+
+class LargeUDF:
+    def __init__(self):
+        self.data = large_object
+
+    def __call__(self, batch):
+        return batch
+
+ds = ray.data.range(1)
+ds = ds.map_batches(LargeUDF, concurrency=1)
+assert ds.take_all() == [{"id": 0}]
+    """
+    output = run_string_as_driver(driver)
```

Comment:
For my own understanding, why do we use `run_string_as_driver` here rather than just directly running the code? We emit the warning on the driver, right?

Reply:
Oh. This wouldn't strictly test if it's printed (that assumes logging is configured correctly), but you can test that it is logged:
```python
def test_read_warning_large_parallelism(ray_start_regular, propagate_logs, caplog):
    with caplog.at_level(logging.WARNING, logger="ray.data.read_api"):
        ray.data.range(5000, override_num_blocks=5000).materialize()
    assert (
        "The requested number of read blocks of 5000 is "
        "more than 4x the number of available CPU slots in the cluster" in caplog.text
    ), caplog.text
```